dj/sql/parsing/ast.py

```diff
@@ -0,0 +1,491 @@
+"""
+Types to represent the DJ AST used as an intermediate representation for DJ operations
+"""
+from abc import ABC, abstractmethod
+from dataclasses import dataclass, field, fields
+from enum import Enum
+from itertools import chain, zip_longest
+from typing import (
+    Callable,
+    Generic,
+    Iterator,
+    List,
+    Optional,
+    Set,
+    Type,
+    TypeVar,
+    Union,
+)
+
+from dj.utils import flatten
+
+
+class Node(ABC):
+    """
+    Base class for all DJ AST nodes.
+
+    DJ nodes are python dataclasses with the following patterns:
+        - Attributes are either
+            - primitives (int, float, str, bool, None)
+            - iterable from (list, tuple, set)
+            - Enum
+            - descendant of `Node`
+        - Attributes starting with '_' are "obfuscated" and are not included in `children`
+
+    """
+
+    _parents: Optional[Set["Node"]]
+
+    @property
+    def parents(self) -> Set["Node"]:
+        """
+        get the parents of the node
+        """
+        try:
+            return self._parents  # type: ignore
+        except AttributeError:
+            self._parents = set()
+            return self._parents
```

Comment:
I think I did that because Node isn't a dataclass so I can't use `set` in `field` like I would for other empty defaults. I think it's bad practice to initialize a set (or any reference type) as a class attribute? I also cannot use an init since dataclasses won't respect it. Here's an example:

```python
class Node(ABC):
    _parents: set = set()
    def __init__(self):
        self._parents = set()
        
@dataclass
class A(Node):
    name: str
        
@dataclass
class B(Node):
    value: int

a=A("yo")
a._parents.add(5)

b=B(6)

b._parents #{5}
```

Reply:
oh I see. I don't have a good suggestion for you then. @betodealmeida may...

Reply:
The example above works if you use [`__post_init__`](https://docs.python.org/3/library/dataclasses.html#post-init-processing) instead:

```python
-class Node(ABC):
      _parents: set = set()

      def __post_init__(self):
          self._parents = set()
```

Reply:
Clever I hadn't thought to use it on the Node class!